system/security/shared/caching.cpp

```diff
@@ -679,11 +660,54 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
     return rc;
 }
 
-SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
+
+void CPermissionsCache::managedFileScopesCacheFillThread()
 {
-    if (!m_secMgr)
-        return SecAccess_Full; // if no security manager, all full access to all scopes
+    DBGLOG("CACHE: managedFileScopeCacheFillThread started");
 
+    // When the thread first starts, do a cache fill with the write lock held.
```

Comment:
It is unlikely, but a thread could still request permissions before this thread actually starts.

You either need a semaphore that is signalled after the first fill which the code that starts the thread waits for, or some similar mechanism to check it has been initialised when permissions are requested.

Reply:
See comments in flush() - similar problem.. If cache not populated, need to ensure threads that are reading permissions wait until ready.

Reply:
The 2 cases which can and probably should be treated differently.
1) Initialization - threads calling for permissions, must ensure the cache has been initialized, they should block if the cache is not yet initialized.

2) flush - I think it's okay for the semantics to mean "will flush" - if so, the flush() call itself needn't perform the flush itself (which will lead to invalid results as it stands). It can instead mean "flush needed", and let the managedFileScopesCacheFillThread thread perform the flush and update.
That will mean that a flush() followed by an immediate call to get perms, will return the previous state, i.e. the update will be in-flight, and won't have updated the cache yet, but I think that's okay.